streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -253,17 +290,168 @@ public void buildAndOptimizeTopology() {
+                                                                                      serialized.keySerde(),
+                                                                                      serialized.valueSerde());
+
+            // re-use parent buildPriority to make sure the single repartition graph node is evaluated before downstream nodes
+            optimizedSingleRepartition.setBuildPriority(keyChangingNode.buildPriority());
+
+            for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : entry.getValue()) {
+
+                final StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNodes().contains(keyChangingNode));
+
+                if (keyChangingNodeChild == null) {
+                    throw new StreamsException(String.format("Found a null keyChangingChild node for %s", repartitionNodeToBeReplaced));
+                }
+
+                LOG.debug("Found the child node of the key changer {} from the repartition {}.", keyChangingNodeChild, repartitionNodeToBeReplaced);
+
+                // need to add children of key-changing node as children of optimized repartition
+                // in order to process records from re-partitioning
+                optimizedSingleRepartition.addChild(keyChangingNodeChild);
+
+                LOG.debug("Removing {} from {}  children {}", keyChangingNodeChild, keyChangingNode, keyChangingNode.children());
+                // now remove children from key-changing node
+                keyChangingNode.removeChild(keyChangingNodeChild);
+
+                // now need to get children of repartition node so we can remove repartition node
+                final Collection<StreamsGraphNode> repartitionNodeToBeReplacedChildren = repartitionNodeToBeReplaced.children();
+                final Collection<StreamsGraphNode> parentsOfRepartitionNodeToBeReplaced = repartitionNodeToBeReplaced.parentNodes();
+
+                for (final StreamsGraphNode repartitionNodeToBeReplacedChild : repartitionNodeToBeReplacedChildren) {
+                    for (final StreamsGraphNode parentNode : parentsOfRepartitionNodeToBeReplaced) {
+                        parentNode.addChild(repartitionNodeToBeReplacedChild);
+                    }
+                }
+
+                for (final StreamsGraphNode parentNode : parentsOfRepartitionNodeToBeReplaced) {
+                    parentNode.removeChild(repartitionNodeToBeReplaced);
+                }
+                repartitionNodeToBeReplaced.clearChildren();
+
+                LOG.debug("Updated node {} children {}", optimizedSingleRepartition, optimizedSingleRepartition.children());
+            }
+
+            keyChangingNode.addChild(optimizedSingleRepartition);
+            keyChangingOperationsToOptimizableRepartitionNodes.remove(entry.getKey());
+        }
+    }
+
+    private void maybeUpdateKeyChangingRepartitionNodeMap() {
```

Comment:
IMHO  it's more straightforward to build the key-changing to optimizing nodes map as is.  

Then before optimizing the topology for repartition nodes, we iterate over the map and "intercept" any key-changing nodes that have a `merge` node as a child.

Since the `merge` node sits between the key-changing node and the repartition node(s) we can use the `merge` node as a proxy for all parent key-changing nodes for the associated repartition node children.

Reply:
@mjsax had a comment above that, for the following case:

```
rekeyed = stream1.map();
merged = rekeyed.merged(stream2);
merged.groupByKey()...
```

We may want to do 1) repartition stream1, 2) then merge with stream2 on the mapped key, and 3) do aggregation without repartition any more. Whereas in the current approach we would 1) merge mapped stream1 with stream2 first, and 2) do repartition, and then 3) do aggregation.

But I think it is not only about optimization efficiency, but also about correctness: Note that merge signature is:

```
KStream<K, V> merge(final KStream<K, V> stream);
```

I.e. the merging KStream should have the same key/value type, so the merging operator itself should be also requiring co-partitioning. This makes me thinking that, we should probably think of `merge` also as a `join` from the optimization point of view, that it requires co-partitioning, and hence if there are key-changing operators, we should enforce a repartitioning before merging.

I know this is not enforced today, which may just be a bug. WDYT?